File: helper-code/tack-and-strat-java/src/main/java/com/tackstrat/ai/SeatAi.java
package com.tackstrat.ai;

import com.tackstrat.model.City;
import com.tackstrat.model.GameSession;
import com.tackstrat.model.HexCoord;
import com.tackstrat.model.Unit;
import com.tackstrat.model.UnitKind;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.Random;

/**
 * Minimal hot-seat opponent: expand production, fight when adjacent, follow routes, otherwise
 * march toward the nearest rival city or unit.
 */
public final class SeatAi {

    private SeatAi() {}

    /**
     * Performs one atomic UI-facing action for the current seat when it is an AI player.
     *
     * @return {@code true} if something changed and another tick may be useful this turn.
     */
    public static boolean tick(GameSession session, Random rng) {
        if (session.isOver()) {
            return false;
        }
        if (!session.currentPlayer().computer()) {
            return false;
        }
        int seat = session.currentPlayer().seat();

        if (clearStuckAutoExplore(session, seat)) {
            return true;
        }
        if (tryFoundCity(session)) {
            return true;
        }
        if (pickCityProduction(session, seat, rng)) {
            return true;
        }
        if (tryAdjacentAttacks(session, seat)) {
            return true;
        }
        if (followAnyQueuedRoute(session, seat)) {
            return true;
        }
        return moveOneUnit(session, seat, rng);
    }

    /**
     * Auto-explore units with MP but no legal step never get moves from the AI loop; turn explore off
     * so normal movement can run on the next tick (avoids spinning forever without ending the turn).
     */
    private static boolean clearStuckAutoExplore(GameSession session, int seat) {
        for (Unit u : session.units()) {
            if (u.ownerSeat() != seat || !u.autoExplore() || u.sleeping() || u.movesRemaining() <= 0) continue;
            if (session.legalMoves(u).isEmpty()) {
                session.setUnitAutoExplore(u.id(), false);
                return true;
            }
        }
        return false;
    }

    private static boolean tryFoundCity(GameSession session) {
        for (Unit u : session.units()) {
            if (u.ownerSeat() != session.currentPlayer().seat()) continue;
            if (u.kind() != UnitKind.SETTLER) continue;
            if (!session.canFoundCity(u)) continue;
            return session.foundCity(u.id()).isPresent();
        }
        return false;
    }

    private static boolean pickCityProduction(GameSession session, int seat, Random rng) {
        List<City> mine = new ArrayList<>();
        for (City c : session.cities()) {
            if (c.ownerSeat() == seat) mine.add(c);
        }
        mine.sort(Comparator.comparingInt(City::id));
        for (City c : mine) {
            if (c.currentBuild() != null) continue;
            UnitKind pick = session.cityCountFor(seat) <= 1 && session.round() <= 6
                    ? UnitKind.SCOUT
                    : UnitKind.WARRIOR;
            if (rng.nextDouble() < 0.22 && session.round() > 8) {
                pick = UnitKind.SETTLER;
            }
            return session.setCityProduction(c.id(), pick);
        }
        return false;
    }

    private static boolean tryAdjacentAttacks(GameSession session, int seat) {
        for (Unit u : session.units()) {
            if (u.ownerSeat() != seat || u.movesRemaining() <= 0) continue;
            if (u.kind().attackStrength() <= 0) continue;
            for (HexCoord n : u.coord().neighbors()) {
                Optional<Unit> foe = session.unitAt(n);
                if (foe.isPresent()
                        && foe.get().ownerSeat() != seat
                        && session.legalAttacks(u).contains(n)) {
                    return session.tryAttack(u.id(), n).isPresent();
                }
            }
        }
        return false;
    }

    private static boolean followAnyQueuedRoute(GameSession session, int seat) {
        List<Unit> mine = new ArrayList<>();
        for (Unit u : session.units()) {
            if (u.ownerSeat() == seat && !session.plannedRouteFor(u.id()).isEmpty()) {
                mine.add(u);
            }
        }
        mine.sort(Comparator.comparingInt(Unit::id));
        for (Unit u : mine) {
            if (session.followPlannedRoute(u.id())) {
                return true;
            }
        }
        return false;
    }

    private static boolean moveOneUnit(GameSession session, int seat, Random rng) {
        List<Unit> movers = new ArrayList<>();
        for (Unit u : session.units()) {
            if (u.ownerSeat() == seat && u.movesRemaining() > 0 && !u.autoExplore() && !u.sleeping()) {
                movers.add(u);
            }
        }
        movers.sort(Comparator.comparingInt(Unit::id));
        for (Unit u : movers) {
            List<HexCoord> legal = session.legalMoves(u);
            if (legal.isEmpty()) continue;
            HexCoord enemyFocus = nearestEnemyFrom(session, seat, u.coord());
            HexCoord dest = pickStepToward(legal, enemyFocus, rng);
            if (session.tryMoveUnit(u.id(), dest)) {
                return true;
            }
        }
        return false;
    }

    private static HexCoord nearestEnemyFrom(GameSession session, int mySeat, HexCoord from) {
        HexCoord best = null;
        int bd = Integer.MAX_VALUE;
        for (City c : session.cities()) {
            if (c.ownerSeat() == mySeat) continue;
            int d = from.distanceTo(c.coord());
            if (d < bd) {
                bd = d;
                best = c.coord();
            }
        }
        for (Unit u : session.units()) {
            if (u.ownerSeat() == mySeat) continue;
            int d = from.distanceTo(u.coord());
            if (d < bd) {
                bd = d;
                best = u.coord();
            }
        }
        return best;
    }

    private static HexCoord pickStepToward(List<HexCoord> legal, HexCoord enemyFocus, Random rng) {
        if (enemyFocus == null) {
            return legal.get(rng.nextInt(legal.size()));
        }
        HexCoord pick = legal.get(0);
        int best = Integer.MAX_VALUE;
        for (HexCoord n : legal) {
            int d = n.distanceTo(enemyFocus);
            if (d < best || (d == best && rng.nextBoolean())) {
                best = d;
                pick = n;
            }
        }
        return pick;
    }
}

File: src/engine/seat_ai.cpp
#include "tack/strat/seat_ai.hpp"

#include <algorithm>
#include <limits>
#include <optional>
#include <vector>

#include "tack/strat/game_session.hpp"
#include "tack/strat/hex.hpp"
#include "tack/strat/unit_kind.hpp"

namespace tack::strat {
namespace {

[[nodiscard]] bool hex_vec_contains(std::vector<HexCoord> const& v, HexCoord n) {
  return std::find(v.begin(), v.end(), n) != v.end();
}

[[nodiscard]] bool clear_stuck_auto_explore(GameSession& session, int seat) {
  for (Unit const& u : session.units()) {
    if (u.owner_seat() != seat || !u.auto_explore() || u.sleeping() || u.moves_remaining() <= 0) {
      continue;
    }
    if (session.legal_moves(u).empty()) {
      session.set_unit_auto_explore(u.id(), false);
      return true;
    }
  }
  return false;
}

[[nodiscard]] bool try_found_city(GameSession& session) {
  for (Unit const& u : session.units()) {
    if (u.owner_seat() != session.current_player().seat) {
      continue;
    }
    if (u.kind() != UnitKind::SETTLER) {
      continue;
    }
    if (!session.can_found_city(u)) {
      continue;
    }
    return session.found_city(u.id()).has_value();
  }
  return false;
}

[[nodiscard]] bool pick_city_production(GameSession& session, int seat, JavaRandom& rng) {
  std::vector<int> mine_ids;
  for (City const& c : session.cities()) {
    if (c.owner_seat() == seat) {
      mine_ids.push_back(c.id());
    }
  }
  std::sort(mine_ids.begin(), mine_ids.end());
  for (int cid : mine_ids) {
    auto co = session.city_by_id(cid);
    if (!co || co->current_build().has_value()) {
      continue;
    }
    UnitKind pick =
        session.city_count_for(seat) <= 1 && session.round() <= 6 ? UnitKind::SCOUT : UnitKind::WARRIOR;
    if (rng.next_double() < 0.22 && session.round() > 8) {
      pick = UnitKind::SETTLER;
    }
    return session.set_city_production(cid, pick);
  }
  return false;
}

[[nodiscard]] bool try_adjacent_attacks(GameSession& session, int seat) {
  for (Unit const& u : session.units()) {
    if (u.owner_seat() != seat || u.moves_remaining() <= 0) {
      continue;
    }
    if (unit_attack_strength(u.kind()) <= 0) {
      continue;
    }
    std::vector<HexCoord> const legal_atk = session.legal_attacks(u);
    for (HexCoord n : u.coord().neighbors()) {
      auto foe = session.unit_at(n);
      if (foe && foe->owner_seat() != seat && hex_vec_contains(legal_atk, n)) {
        return session.try_attack(u.id(), n).has_value();
      }
    }
  }
  return false;
}

[[nodiscard]] bool follow_any_queued_route(GameSession& session, int seat) {
  std::vector<int> mine_ids;
  for (Unit const& u : session.units()) {
    if (u.owner_seat() == seat && !session.planned_route_for(u.id()).empty()) {
      mine_ids.push_back(u.id());
    }
  }
  std::sort(mine_ids.begin(), mine_ids.end());
  for (int uid : mine_ids) {
    if (session.follow_planned_route(uid)) {
      return true;
    }
  }
  return false;
}

[[nodiscard]] std::optional<HexCoord> nearest_enemy_from(GameSession const& session, int my_seat,
                                                         HexCoord from) {
  std::optional<HexCoord> best;
  int bd = std::numeric_limits<int>::max();
  for (City const& c : session.cities()) {
    if (c.owner_seat() == my_seat) {
      continue;
    }
    int d = from.distance_to(c.coord());
    if (d < bd) {
      bd = d;
      best = c.coord();
    }
  }
  for (Unit const& u : session.units()) {
    if (u.owner_seat() == my_seat) {
      continue;
    }
    int d = from.distance_to(u.coord());
    if (d < bd) {
      bd = d;
      best = u.coord();
    }
  }
  return best;
}

[[nodiscard]] HexCoord pick_step_toward(std::vector<HexCoord> const& legal,
                                        std::optional<HexCoord> enemy_focus, JavaRandom& rng) {
  if (!enemy_focus.has_value()) {
    return legal[static_cast<std::size_t>(rng.next_int_bounded(static_cast<int>(legal.size())))];
  }
  HexCoord pick = legal.front();
  int best = std::numeric_limits<int>::max();
  for (HexCoord n : legal) {
    int d = n.distance_to(*enemy_focus);
    if (d < best || (d == best && rng.next_bool())) {
      best = d;
      pick = n;
    }
  }
  return pick;
}

[[nodiscard]] bool move_one_unit(GameSession& session, int seat, JavaRandom& rng) {
  std::vector<int> mover_ids;
  for (Unit const& u : session.units()) {
    if (u.owner_seat() == seat && u.moves_remaining() > 0 && !u.auto_explore() && !u.sleeping()) {
      mover_ids.push_back(u.id());
    }
  }
  std::sort(mover_ids.begin(), mover_ids.end());
  for (int uid : mover_ids) {
    auto uo = session.unit_by_id(uid);
    if (!uo) {
      continue;
    }
    Unit const& u = *uo;
    std::vector<HexCoord> legal = session.legal_moves(u);
    if (legal.empty()) {
      continue;
    }
    std::optional<HexCoord> focus = nearest_enemy_from(session, seat, u.coord());
    HexCoord dest = pick_step_toward(legal, focus, rng);
    if (session.try_move_unit(uid, dest)) {
      return true;
    }
  }
  return false;
}

}  // namespace

JavaRandom seat_ai_rng(GameSession const& session) {
  std::int64_t const salt =
      session.world_seed() ^ (static_cast<std::int64_t>(session.round()) << 16) ^
      (static_cast<std::int64_t>(session.current_player_index()) << 24);
  return JavaRandom(salt);
}

bool seat_ai_tick(GameSession& session, JavaRandom& rng) {
  if (session.is_over()) {
    return false;
  }
  if (!session.current_player().computer) {
    return false;
  }
  int const seat = session.current_player().seat;

  if (clear_stuck_auto_explore(session, seat)) {
    return true;
  }
  if (try_found_city(session)) {
    return true;
  }
  if (pick_city_production(session, seat, rng)) {
    return true;
  }
  if (try_adjacent_attacks(session, seat)) {
    return true;
  }
  if (follow_any_queued_route(session, seat)) {
    return true;
  }
  return move_one_unit(session, seat, rng);
}

}  // namespace tack::strat
